csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/WeakObjectManager.cs

```diff
@@ -0,0 +1,225 @@
+
+            int garbageCount;
+            var aliveList = ReleseGarbageCollectedObjects(endTime, out garbageCount);
+
+            var timeReleaseGarbage = DateTime.Now;
+            aliveList.ForEach(item => weakReferences.Enqueue(item));
+            var timeStoreAlive = DateTime.Now;
+
+            logger.LogInfo("check end : released {0} garbage, remain {1} alive, used {2} ms : release garbage used {3} ms, store alive used {4} ms",
+                    garbageCount, weakReferences.Count, (DateTime.Now - beginTime).TotalMilliseconds,
+                    (timeReleaseGarbage - beginTime).TotalMilliseconds,
+                    (timeStoreAlive - timeReleaseGarbage).TotalMilliseconds
+                );
+        }
+
+        private List<WeakReferenceObjectIdPair> ReleseGarbageCollectedObjects(DateTime endTime, out int garbageCount)
+        {
+            var aliveList = new List<WeakReferenceObjectIdPair>();
+            garbageCount = 0;
+            WeakReferenceObjectIdPair weakReferenceObjectIdPair;
+
+            while (weakReferences.TryDequeue(out weakReferenceObjectIdPair))
+            {
+                var weakRef = weakReferenceObjectIdPair.Key;
+                if (weakRef.IsAlive)
+                {
+                    aliveList.Add(weakReferenceObjectIdPair);
+                }
+                else
+                {
+                    objectReleaser.ReleaseObject(weakReferenceObjectIdPair.Value);
+                    garbageCount++;
+                }
+
+                if (DateTime.Now > endTime)
+                {
+                    logger.LogDebug("Stop releasing as exceeded allowed time : {0}", endTime.ToString("yyyy-MM-dd HH:mm:ss.fff"));
+                    break;
+                }
+            }
+
+            return aliveList;
+        }
+
+        /// <summary>
+        /// It can be an expensive operation. ** Do not use ** unless there is a real need for this method
+        /// </summary>
+        /// <returns></returns>
+        public long GetAliveCount()
+        {
+            //copying to get alive count at the time of this method call
+            var copiedList = new Queue<WeakReferenceObjectIdPair>(weakReferences);
+            var count = 0;
+            foreach (var weakReference in copiedList)
+            {
+                if (weakReference.Key.IsAlive)
+                {
+                    count++;
+                }
+            }
+
+            return count;
+        }
+
+        public virtual void Dispose()
+        {
+            logger.LogInfo("Dispose {0}", this.GetType());
+            shouldKeepRunning = false;
+        }
+    }
+
+    internal interface IObjectReleaser
+    {
+        void ReleaseObject(string objId);
+    }
+
+    internal class JvmObjectReleaser : IObjectReleaser
+    {
+        private const string ReleaseHandler = "SparkCLRHandler";
+        private const string ReleaseMethod = "rm";
```

Comment:
@skaarthik : I've an idea as following, how about your thinking?
To test it, need to call [CSharpBackendHandler.handleMethodCall ](https://github.com/Microsoft/Mobius/blob/master/scala/src/main/org/apache/spark/api/csharp/CSharpBackendHandler.scala#L114)
the main trivial is about the **DataInputStream** of the parameters which serialized from C# [PayloadHelper.BuildPayload](https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/JvmBridge.cs#L74) related with [SerDe.cs](https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/SerDe.cs) and [PayloadHelper.cs](https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/PayloadHelper.cs) 

I've 3 ways to do this:
1.Just make the byte array by C# as hard code in Scala unit test.
2.Call [CSharpBackendHandler.handleMethodCall ](https://github.com/Microsoft/Mobius/blob/master/scala/src/main/org/apache/spark/api/csharp/CSharpBackendHandler.scala#L114) in C# Adapter.dll from Scala to build it.
3.Clone and translate the [SerDe.cs](https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/SerDe.cs) and [PayloadHelper.cs](https://github.com/Microsoft/Mobius/blob/master/csharp/Adapter/Microsoft.Spark.CSharp/Interop/Ipc/PayloadHelper.cs) to Scala.

As long as the SparkCLR interfaces and protocols in C# and Scala not changed, all of them works the same. 
What about your opinion ? or any solution beyond ?